fix dynamo tracking numpy 2 ops

haifeng-jin opened the pull request:
# Summary

Support `torch.compile()` to trace through `numpy.random` ops when used with NumPy 2.

# Details:

Fixes #136559
As we upgrade to NumPy 2, torch falsely filtered out `numpy.random` as unsupported in dynamo tracking.
This PR changes the filtering rules to include them while keeping behavior with numpy 1 unchanged.


Before this PR, the following tests failed:

```
PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/dynamo/test_functions.py -k FunctionTests.test_numpy_random
PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/dynamo/test_unspec.py -k UnspecTests.test_to_tensor
PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/test_fake_tensor.py -k FakeTensorTest.test_export_numpy
PYTORCH_TEST_WITH_ASAN=1 PYTORCH_TEST_WITH_UBSAN=1 python test/test_fake_tensor.py -k PropagateRealTensorsFakeTensorTest.test_export_numpy_propagate_real_tensors
```

With this PR, the supported/unsupported ops in NumPy 1 are not changed.
For NumPy 2, only the `numpy.random` ops that are already supported with NumPy 1 are added to the supported list.

I used the following scripts to check the differences before and after the change for both NumPy 1 & 2.
The output is empty for NumPy 1 since there is no change.
The output is a list of `numpy.random` that considered supported for NumPy 2.

```py
from torch._dynamo import trace_rules
import numpy as np


def new_numpy_function_ids():
    unsupported_funcs = {"seed", "ranf", "get_bit_generator", "RandomState", "set_bit_generator", "sample"}

    def is_supported(k, v, mod):
        if not callable(v):
            return False
        if not getattr(v, "__module__", None):
            return True
        if v.__module__ == mod.__name__:
            return True
        if v.__module__ == "numpy.random.mtrand" and mod.__name__== "numpy.random" and k not in unsupported_funcs:
            return True
        return False
    rv = {}
    for mod in trace_rules.NP_SUPPORTED_MODULES:
        for k, v in mod.__dict__.items():
            if is_supported(k, v, mod):
                rv[id(v)] = f"{mod.__name__}.{k}"
    return rv

def old_numpy_function_ids():
    rv = {}
    for mod in trace_rules.NP_SUPPORTED_MODULES:
        rv.update(
            {
                id(v): f"{mod.__name__}.{k}"
                for k, v in mod.__dict__.items()
                if callable(v)
                and (getattr(v, "__module__", None) or mod.__name__) == mod.__name__
            }
        )
    return rv

rv1 = set(old_numpy_function_ids().values())
rv2 = set(new_numpy_function_ids().values())

for v in (rv1 - rv2):
    print(v)
print("****")
for v in (rv2 - rv1):
    print(v)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @kiukchung @lezcano

haifeng-jin committed:
fix dynamo tracking numpy 2 ops

zou3519 commented:
I'm not the right person to review this, @anijain2305 @williamwen42 do one of you want to take this?

haifeng-jin commented:
Can we also request a review from @lezcano?
Thanks!

williamwen42 commented on torch/_dynamo/trace_rules.py:
```diff
@@ -3021,16 +3021,35 @@ def _polyfilled_function_ids() -> Set[int]:
 
 @FunctionIdSet
 def _numpy_function_ids() -> Dict[int, str]:
+    unsupported_funcs = {
```
You might not need this set - if we actually encounter an unsupported numpy function, I believe we would just graph break when attempting to trace the call in dynamo. You can try removing this set and seeing if the tests pass.

haifeng-jin replied:
I have update the PR.
I did some quick local tests. They passed.
Thanks!

haifeng-jin committed:
remove the unnecessary list of unsupported numpy ops

williamwen42 commented:
@pytorchbot rebase

williamwen42 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but I am seeing inductor failure with hf_BigBird number of graph breaks after it lands' -c nosignal

huydhn commented:
Maybe the the expected value just needs to be updated https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo/ci_expected_accuracy

haifeng-jin commented:
@huydhn Thank you!
Would you mind  share how should I reproduce the errors that caused by the PR?
like the command for running the failed tests.

huydhn commented:
Here is an example failed job [GH job link](https://github.com/pytorch/pytorch/actions/runs/11637726618/job/32412391920) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/77b72d686e3bc317733db41beef1638712bfdd2c)

I have added `ciflow/inductor` in your PR, so you can see they are failing on the PR too https://github.com/pytorch/pytorch/pull/138686#issuecomment-2431003929

haifeng-jin committed:
Exclude the unsupported ops.

This reverts commit 3b87731350752f2c3c16543fa57ff51fe22a54b5.

haifeng-jin commented:
I have locally reproduced the Hugging face BigBird graph breaks and fixed it.
See more about the cause and fix in [this thread](https://github.com/pytorch/pytorch/pull/138686#discussion_r1821625411).

<details>

<summary>More details on reproducing it locally</summary>

I cloned the torch benchmark repo and followed the installation instructions.
Added it to the Python path and run the following command in the PyTorch root repo.

```shell
python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --inductor --device cpu --inference --float32 --total-partitions 2 --partition-id 0 --output inference_torchbench.csv -k hf_BigBird
```

</details>

haifeng-jin commented:
@williamwen42 @huydhn Would you please help review and merge the PR? Thanks!

williamwen42 commented:
@pytorchbot merge

haifeng-jin commented:
@pytorchbot label "topic: bug fixes"

haifeng-jin commented:
@pytorchbot label "release notes: dynamo"
